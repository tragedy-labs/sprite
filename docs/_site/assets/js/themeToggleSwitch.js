function initializeThemeToggle() {
  const element = document.getElementById("theme_toggle");
  const theme = localStorage.getItem("theme");

  const prefersTheme = (theme) =>
    window.matchMedia(`(prefers-color-scheme: ${theme})`).matches;

  const setTheme = (theme) => {
    localStorage.setItem("theme", theme);
    document.body.className = theme;
    setSwitchState(theme);
  };

  const setSwitchState = (theme) => {
    switch (theme) {
      case "dark":
        element.checked = true;
        break;
      case "light":
        element.checked = false;
        break;
      default:
        throw new Error("unknown theme");
    }
  };

  const determinePosition = () => {
    if (theme) {
      setTheme(theme);
    } else {
      if (prefersTheme("dark")) {
        setTheme("dark");
        return;
      } else {
        setTheme("light");
      }
    }
  };

  element.addEventListener("click", () => {
    if (element.checked) {
      setTheme("dark");
    } else {
      setTheme("light");
    }
  });

  determinePosition();
}

document.addEventListener("DOMContentLoaded", initializeThemeToggle);
