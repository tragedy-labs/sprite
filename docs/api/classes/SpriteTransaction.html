<!DOCTYPE html><html class="default" lang="en"><head><meta charSet="utf-8"/><meta http-equiv="x-ua-compatible" content="IE=edge"/><title>SpriteTransaction | @tragedy-labs/sprite</title><meta name="description" content="Documentation for @tragedy-labs/sprite"/><meta name="viewport" content="width=device-width, initial-scale=1"/><link rel="stylesheet" href="../assets/style.css"/><link rel="stylesheet" href="../assets/highlight.css"/><script defer src="../assets/main.js"></script><script async src="../assets/icons.js" id="tsd-icons-script"></script><script async src="../assets/search.js" id="tsd-search-script"></script><script async src="../assets/navigation.js" id="tsd-nav-script"></script></head><body><script>document.documentElement.dataset.theme = localStorage.getItem("tsd-theme") || "os";document.body.style.display="none";setTimeout(() => app?app.showPage():document.body.style.removeProperty("display"),500)</script><header class="tsd-page-toolbar"><div class="tsd-toolbar-contents container"><div class="table-cell" id="tsd-search" data-base=".."><div class="field"><label for="tsd-search-field" class="tsd-widget tsd-toolbar-icon search no-caption"><svg width="16" height="16" viewBox="0 0 16 16" fill="none"><use href="../assets/icons.svg#icon-search"></use></svg></label><input type="text" id="tsd-search-field" aria-label="Search"/></div><div class="field"><div id="tsd-toolbar-links"></div></div><ul class="results"><li class="state loading">Preparing search index...</li><li class="state failure">The search index is not available</li></ul><a href="../index.html" class="title">@tragedy-labs/sprite</a></div><div class="table-cell" id="tsd-widgets"><a href="#" class="tsd-widget tsd-toolbar-icon menu no-caption" data-toggle="menu" aria-label="Menu"><svg width="16" height="16" viewBox="0 0 16 16" fill="none"><use href="../assets/icons.svg#icon-menu"></use></svg></a></div></div></header><div class="container container-main"><div class="col-content"><div class="tsd-page-title"><ul class="tsd-breadcrumb"><li><a href="../modules.html">@tragedy-labs/sprite</a></li><li><a href="SpriteTransaction.html">SpriteTransaction</a></li></ul><h1>Class SpriteTransaction</h1></div><section class="tsd-panel tsd-comment"><div class="tsd-comment tsd-typography"><p>A transaction in Sprite, contains the transaction id, and methods to
commit or rollback the transaction.</p>
</div><div class="tsd-comment tsd-typography"><h4>Param: session</h4><p>the <code>DatabaseSession</code> the transaction is taking place of</p>
<h4>Param: transactionId</h4><p>the id of the transaction</p>
</div></section><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L11">src/transaction/SpriteTransaction.ts:11</a></li></ul></aside><section class="tsd-panel-group tsd-index-group"><section class="tsd-panel tsd-index-panel"><details class="tsd-index-content tsd-index-accordion" open><summary class="tsd-accordion-summary tsd-index-summary"><h5 class="tsd-index-heading uppercase" role="button" aria-expanded="false" tabIndex="0"><svg width="16" height="16" viewBox="0 0 16 16" fill="none"><use href="../assets/icons.svg#icon-chevronSmall"></use></svg> Index</h5></summary><div class="tsd-accordion-details"><section class="tsd-index-section"><h3 class="tsd-index-heading">Constructors</h3><div class="tsd-index-list"><a href="SpriteTransaction.html#constructor" class="tsd-index-link"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-512"></use></svg><span>constructor</span></a>
</div></section><section class="tsd-index-section"><h3 class="tsd-index-heading">Properties</h3><div class="tsd-index-list"><a href="SpriteTransaction.html#_committed" class="tsd-index-link tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_committed</span></a>
<a href="SpriteTransaction.html#_id" class="tsd-index-link tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_id</span></a>
<a href="SpriteTransaction.html#_rolledBack" class="tsd-index-link tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_rolled<wbr/>Back</span></a>
<a href="SpriteTransaction.html#_session" class="tsd-index-link tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_session</span></a>
</div></section><section class="tsd-index-section"><h3 class="tsd-index-heading">Accessors</h3><div class="tsd-index-list"><a href="SpriteTransaction.html#committed" class="tsd-index-link"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-262144"></use></svg><span>committed</span></a>
<a href="SpriteTransaction.html#id" class="tsd-index-link"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-262144"></use></svg><span>id</span></a>
<a href="SpriteTransaction.html#rolledBack" class="tsd-index-link"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-262144"></use></svg><span>rolled<wbr/>Back</span></a>
</div></section><section class="tsd-index-section"><h3 class="tsd-index-heading">Methods</h3><div class="tsd-index-list"><a href="SpriteTransaction.html#commit" class="tsd-index-link"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-2048"></use></svg><span>commit</span></a>
<a href="SpriteTransaction.html#crud" class="tsd-index-link"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-2048"></use></svg><span>crud</span></a>
<a href="SpriteTransaction.html#rollback" class="tsd-index-link"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-2048"></use></svg><span>rollback</span></a>
</div></section></div></details></section></section><section class="tsd-panel-group tsd-member-group"><h2>Constructors</h2><section class="tsd-panel tsd-member"><a id="constructor" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><span>constructor</span><a href="#constructor" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><ul class="tsd-signatures"><li class="tsd-signature tsd-anchor-link"><a id="constructor.new_SpriteTransaction" class="tsd-anchor"></a><span class="tsd-kind-constructor-signature">new <wbr/>Sprite<wbr/>Transaction</span><span class="tsd-signature-symbol">(</span><span class="tsd-kind-parameter">session</span>, <span class="tsd-kind-parameter">transactionId</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><a href="SpriteTransaction.html" class="tsd-signature-type tsd-kind-class">SpriteTransaction</a><a href="#constructor.new_SpriteTransaction" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></li><li class="tsd-description"><div class="tsd-parameters"><h4 class="tsd-parameters-title">Parameters</h4><ul class="tsd-parameter-list"><li><span><span class="tsd-kind-parameter">session</span>: <span class="tsd-signature-type">DatabaseSession</span></span></li><li><span><span class="tsd-kind-parameter">transactionId</span>: <span class="tsd-signature-type">string</span></span></li></ul></div><h4 class="tsd-returns-title">Returns <a href="SpriteTransaction.html" class="tsd-signature-type tsd-kind-class">SpriteTransaction</a></h4><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L17">src/transaction/SpriteTransaction.ts:17</a></li></ul></aside></li></ul></section></section><section class="tsd-panel-group tsd-member-group"><h2>Properties</h2><section class="tsd-panel tsd-member tsd-is-private"><a id="_committed" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><code class="tsd-tag ts-flagPrivate">Private</code> <span>_committed</span><a href="#_committed" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><div class="tsd-signature"><span class="tsd-kind-property">_committed</span><span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol"> = false</span></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L16">src/transaction/SpriteTransaction.ts:16</a></li></ul></aside></section><section class="tsd-panel tsd-member tsd-is-private"><a id="_id" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><code class="tsd-tag ts-flagPrivate">Private</code> <span>_id</span><a href="#_id" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><div class="tsd-signature"><span class="tsd-kind-property">_id</span><span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">string</span></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L13">src/transaction/SpriteTransaction.ts:13</a></li></ul></aside></section><section class="tsd-panel tsd-member tsd-is-private"><a id="_rolledBack" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><code class="tsd-tag ts-flagPrivate">Private</code> <span>_rolled<wbr/>Back</span><a href="#_rolledBack" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><div class="tsd-signature"><span class="tsd-kind-property">_rolled<wbr/>Back</span><span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol"> = false</span></div><div class="tsd-comment tsd-typography"><p>State</p>
</div><div class="tsd-comment tsd-typography"></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L15">src/transaction/SpriteTransaction.ts:15</a></li></ul></aside></section><section class="tsd-panel tsd-member tsd-is-private"><a id="_session" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><code class="tsd-tag ts-flagPrivate">Private</code> <span>_session</span><a href="#_session" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><div class="tsd-signature"><span class="tsd-kind-property">_session</span><span class="tsd-signature-symbol">:</span> <span class="tsd-signature-type">DatabaseSession</span></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L12">src/transaction/SpriteTransaction.ts:12</a></li></ul></aside></section></section><section class="tsd-panel-group tsd-member-group"><h2>Accessors</h2><section class="tsd-panel tsd-member"><a id="committed" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><span>committed</span><a href="#committed" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><ul class="tsd-signatures"><li class="tsd-signature" id="committed.committed-1"><span class="tsd-signature-keyword">get</span> committed<span class="tsd-signature-symbol">(</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">boolean</span></li><li class="tsd-description"><div class="tsd-comment tsd-typography"><p>Whether or not the transaction has been committed</p>
</div><h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">boolean</span></h4><div class="tsd-comment tsd-typography"></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L30">src/transaction/SpriteTransaction.ts:30</a></li></ul></aside></li></ul></section><section class="tsd-panel tsd-member"><a id="id" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><span>id</span><a href="#id" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><ul class="tsd-signatures"><li class="tsd-signature" id="id.id-1"><span class="tsd-signature-keyword">get</span> id<span class="tsd-signature-symbol">(</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">string</span></li><li class="tsd-description"><div class="tsd-comment tsd-typography"><p>The trasaction ID</p>
</div><h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">string</span></h4><div class="tsd-comment tsd-typography"></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L22">src/transaction/SpriteTransaction.ts:22</a></li></ul></aside></li></ul></section><section class="tsd-panel tsd-member"><a id="rolledBack" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><span>rolled<wbr/>Back</span><a href="#rolledBack" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><ul class="tsd-signatures"><li class="tsd-signature" id="rolledBack.rolledBack-1"><span class="tsd-signature-keyword">get</span> rolledBack<span class="tsd-signature-symbol">(</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">boolean</span></li><li class="tsd-description"><div class="tsd-comment tsd-typography"><p>Whether or not the transaction has been rolledback</p>
</div><h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">boolean</span></h4><div class="tsd-comment tsd-typography"></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L26">src/transaction/SpriteTransaction.ts:26</a></li></ul></aside></li></ul></section></section><section class="tsd-panel-group tsd-member-group"><h2>Methods</h2><section class="tsd-panel tsd-member"><a id="commit" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><span>commit</span><a href="#commit" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><ul class="tsd-signatures"><li class="tsd-signature tsd-anchor-link"><a id="commit.commit-1" class="tsd-anchor"></a><span class="tsd-kind-call-signature">commit</span><span class="tsd-signature-symbol">(</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol">&gt;</span><a href="#commit.commit-1" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></li><li class="tsd-description"><div class="tsd-comment tsd-typography"><p>Commit the transaction.</p>
</div><h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol">&gt;</span></h4><p><code>true</code> if the transaction was commited.</p>
<div class="tsd-comment tsd-typography"></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L37">src/transaction/SpriteTransaction.ts:37</a></li></ul></aside></li></ul></section><section class="tsd-panel tsd-member"><a id="crud" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><span>crud</span><a href="#crud" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><ul class="tsd-signatures"><li class="tsd-signature tsd-anchor-link"><a id="crud.crud-1" class="tsd-anchor"></a><span class="tsd-kind-call-signature">crud</span><span class="tsd-signature-symbol">&lt;</span><a class="tsd-signature-type tsd-kind-type-parameter" href="SpriteTransaction.html#crud.crud-1.T">T</a><span class="tsd-signature-symbol">&gt;</span><span class="tsd-signature-symbol">(</span><span class="tsd-kind-parameter">language</span>, <span class="tsd-kind-parameter">command</span>, <span class="tsd-kind-parameter">parameters</span><span class="tsd-signature-symbol">?</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><a class="tsd-signature-type tsd-kind-type-parameter" href="SpriteTransaction.html#crud.crud-1.T">T</a><span class="tsd-signature-symbol">&gt;</span><a href="#crud.crud-1" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></li><li class="tsd-description"><div class="tsd-comment tsd-typography"><p>Perform a CRUD operation in the transaction.</p>
</div><section class="tsd-panel"><h4>Type Parameters</h4><ul class="tsd-type-parameter-list"><li><span><a id="crud.crud-1.T" class="tsd-anchor"></a><span class="tsd-kind-type-parameter">T</span></span></li></ul></section><div class="tsd-parameters"><h4 class="tsd-parameters-title">Parameters</h4><ul class="tsd-parameter-list"><li><span><span class="tsd-kind-parameter">language</span>: <span class="tsd-signature-type">ArcadeSupportedQueryLanguages</span></span><div class="tsd-comment tsd-typography"><p>the query language to use</p>
</div><div class="tsd-comment tsd-typography"></div></li><li><span><span class="tsd-kind-parameter">command</span>: <span class="tsd-signature-type">string</span></span><div class="tsd-comment tsd-typography"><p>the command to execute</p>
</div><div class="tsd-comment tsd-typography"></div></li><li><span><code class="tsd-tag ts-flagOptional">Optional</code> <span class="tsd-kind-parameter">parameters</span>: <span class="tsd-signature-type">Record</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol">, </span><span class="tsd-signature-type">string</span><span class="tsd-signature-symbol"> | </span><span class="tsd-signature-type">number</span><span class="tsd-signature-symbol"> | </span><span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol">&gt;</span></span></li></ul></div><h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><a class="tsd-signature-type tsd-kind-type-parameter" href="SpriteTransaction.html#crud.crud-1.T">T</a><span class="tsd-signature-symbol">&gt;</span></h4><p>the result of the CRUD operation</p>
<div class="tsd-comment tsd-typography"></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L56">src/transaction/SpriteTransaction.ts:56</a></li></ul></aside></li></ul></section><section class="tsd-panel tsd-member"><a id="rollback" class="tsd-anchor"></a><h3 class="tsd-anchor-link"><span>rollback</span><a href="#rollback" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></h3><ul class="tsd-signatures"><li class="tsd-signature tsd-anchor-link"><a id="rollback.rollback-1" class="tsd-anchor"></a><span class="tsd-kind-call-signature">rollback</span><span class="tsd-signature-symbol">(</span><span class="tsd-signature-symbol">)</span><span class="tsd-signature-symbol">: </span><span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol">&gt;</span><a href="#rollback.rollback-1" aria-label="Permalink" class="tsd-anchor-icon"><svg viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-anchor"></use></svg></a></li><li class="tsd-description"><div class="tsd-comment tsd-typography"><p>Rollback the transaction.</p>
</div><h4 class="tsd-returns-title">Returns <span class="tsd-signature-type">Promise</span><span class="tsd-signature-symbol">&lt;</span><span class="tsd-signature-type">boolean</span><span class="tsd-signature-symbol">&gt;</span></h4><p><code>true</code> if the transaction was commited.</p>
<div class="tsd-comment tsd-typography"></div><aside class="tsd-sources"><ul><li>Defined in <a href="https://github.com/tragedy-labs/sprite/blob/fe1f8c244b966e03ec5257436d6e5f518e57f3f3/src/transaction/SpriteTransaction.ts#L45">src/transaction/SpriteTransaction.ts:45</a></li></ul></aside></li></ul></section></section></div><div class="col-sidebar"><div class="page-menu"><div class="tsd-navigation settings"><details class="tsd-index-accordion"><summary class="tsd-accordion-summary"><h3><svg width="20" height="20" viewBox="0 0 24 24" fill="none"><use href="../assets/icons.svg#icon-chevronDown"></use></svg>Settings</h3></summary><div class="tsd-accordion-details"><div class="tsd-filter-visibility"><h4 class="uppercase">Member Visibility</h4><form><ul id="tsd-filter-options"><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-protected" name="protected"/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>Protected</span></label></li><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-private" name="private"/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>Private</span></label></li><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-inherited" name="inherited" checked/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>Inherited</span></label></li><li class="tsd-filter-item"><label class="tsd-filter-input"><input type="checkbox" id="tsd-filter-external" name="external"/><svg width="32" height="32" viewBox="0 0 32 32" aria-hidden="true"><rect class="tsd-checkbox-background" width="30" height="30" x="1" y="1" rx="6" fill="none"></rect><path class="tsd-checkbox-checkmark" d="M8.35422 16.8214L13.2143 21.75L24.6458 10.25" stroke="none" stroke-width="3.5" stroke-linejoin="round" fill="none"></path></svg><span>External</span></label></li></ul></form></div><div class="tsd-theme-toggle"><h4 class="uppercase">Theme</h4><select id="tsd-theme"><option value="os">OS</option><option value="light">Light</option><option value="dark">Dark</option></select></div></div></details></div><details open class="tsd-index-accordion tsd-page-navigation"><summary class="tsd-accordion-summary"><h3><svg width="20" height="20" viewBox="0 0 24 24" fill="none"><use href="../assets/icons.svg#icon-chevronDown"></use></svg>On This Page</h3></summary><div class="tsd-accordion-details"><a href="#constructor" class=""><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-512"></use></svg><span>constructor</span></a><a href="#_committed" class="tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_committed</span></a><a href="#_id" class="tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_id</span></a><a href="#_rolledBack" class="tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_rolled<wbr/>Back</span></a><a href="#_session" class="tsd-is-private"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1024"></use></svg><span>_session</span></a><a href="#committed" class=""><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-262144"></use></svg><span>committed</span></a><a href="#id" class=""><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-262144"></use></svg><span>id</span></a><a href="#rolledBack" class=""><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-262144"></use></svg><span>rolled<wbr/>Back</span></a><a href="#commit" class=""><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-2048"></use></svg><span>commit</span></a><a href="#crud" class=""><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-2048"></use></svg><span>crud</span></a><a href="#rollback" class=""><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-2048"></use></svg><span>rollback</span></a></div></details></div><div class="site-menu"><nav class="tsd-navigation"><a href="../modules.html"><svg class="tsd-kind-icon" viewBox="0 0 24 24"><use href="../assets/icons.svg#icon-1"></use></svg><span>@tragedy-labs/sprite</span></a><ul class="tsd-small-nested-navigation" id="tsd-nav-container" data-base=".."><li>Loading...</li></ul></nav></div></div></div><footer><p class="tsd-generator">Generated using <a href="https://typedoc.org/" target="_blank">TypeDoc</a></p></footer><div class="overlay"></div></body></html>