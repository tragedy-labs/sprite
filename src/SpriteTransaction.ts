import { SpriteDatabase } from './SpriteDatabase.js';

export class SpriteTransaction {
  private _id: string;
  private _committed: boolean = false;
  private database: SpriteDatabase;
  constructor(database: SpriteDatabase, transactionId: string) {
    this.database = database;
    this._id = transactionId;
  }
  /** The trasaction ID */
  get id() {
    return this._id;
  }
  /** Whether or not the transaction has been committed */
  get committed() {
    return this._committed;
  }
  commit = async () => {
    // TODO: This old committed logic is flawed and perhaps replacing
    // with a status would be a better approach.
    this._committed = await this.database.commitTransaction(this.id);
    return this.committed;
  };
  rollback = async () => this.database.rollbackTransaction(this.id);
}
