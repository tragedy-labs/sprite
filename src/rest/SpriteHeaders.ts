import { SpriteTransaction } from '../transaction/SpriteTransaction.js';
import { DatabaseSession } from '../session/DatabaseSession.js';
import { Auth } from './Auth.js';
import { ServerSession } from '@/session/ServerSession.js';
import { ArcadeValidation } from '@/validation/ArcadeValidation.js';

export enum HeaderKeys {
  ContentType = 'Content-Type',
  Authorization = 'Authorization',
  ArcadeSessionId = 'arcadedb-session-id'
}

export type SpriteBasicHeadersInit = HeadersInit & {
  [HeaderKeys.Authorization]: string;
  [HeaderKeys.ContentType]: 'application/json';
};

export type SpriteHeadersInit = SpriteBasicHeadersInit & {
  [HeaderKeys.ArcadeSessionId]?: string;
};

class SpriteHeaders {
  public static compose(
    session: DatabaseSession | ServerSession,
    transaction?: SpriteTransaction
  ): SpriteHeadersInit {
    try {
      if (transaction) {
        ArcadeValidation.transaction(transaction);
        return {
          [HeaderKeys.ContentType]: session.headers[HeaderKeys.ContentType],
          [HeaderKeys.Authorization]: session.headers[HeaderKeys.Authorization],
          [HeaderKeys.ArcadeSessionId]: transaction.id
        };
      } else {
        return session.headers;
      }
    } catch (error) {
      throw new Error(
        'Could not compose REST headers. Typically this is because a valid transaction object was not supplied.',
        { cause: error }
      );
    }
  }
  public static initialize = (
    username: string,
    password: string
  ): SpriteBasicHeadersInit => ({
    [HeaderKeys.ContentType]: 'application/json',
    [HeaderKeys.Authorization]: Auth.basic(username, password)
  });
}

export { SpriteHeaders };
