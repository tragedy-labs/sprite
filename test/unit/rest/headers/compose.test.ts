// Lib
import { SpriteHeaders } from '@/rest/SpriteHeaders.js';
import { SpriteTransaction } from '@/transaction/SpriteTransaction.js';

// Testing
import {
  testTransaction as TRX,
  TestServerSession as SESSION,
  headersWithTransaction as HEADERS
} from '@test/variables.js';

describe('SpriteHeaders.compose', () => {
  it('should compose headers with transaction id', () => {
    // Act
    const headers = SpriteHeaders.compose(SESSION, TRX);
    // Assert
    expect(headers).toEqual(HEADERS);
  });

  it('should throw an error if it recieves an invalid transaction', () => {
    // Arrange
    const invalidTransaction = {
      invalid: 'object'
    } as unknown as SpriteTransaction;
    // Assert
    expect(() =>
      SpriteHeaders.compose(SESSION, invalidTransaction)
    ).toThrowErrorMatchingSnapshot();
  });
});
