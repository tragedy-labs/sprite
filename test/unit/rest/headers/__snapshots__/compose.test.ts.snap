// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpriteHeaders.compose should throw an error if it recieves an invalid transaction 1`] = `"Could not compose REST headers. Typically this is because a valid transaction object was not supplied."`;
